### PR TITLE
Enable HSwish - ngraph op, fusion/decomposition and reference implementation

### PR DESCRIPTION
**Pull request by mitruska**
Enable [HSwish ](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/activation/HSwish_4.md)(new in opset4) operator 

- [x]  ngraph op
- [x]  ngraph python API
- [x]  reference implementation
- [x] fusion transformation pass
- [x] decomposition transformation pass

### CODE REVIEW

File: ngraph/core/include/ngraph/runtime/reference/hswish.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        namespace reference
+        {
+            template <typename T>
+            void hswish(const T* arg, T* out, size_t count)
+            {
+                for (size_t i = 0; i < count; i++)
+                {
+                    out[i] = arg[i] * std::min<T>(std::max<T>(arg[i] + 3.0f, 0.0f), 6.0f) *
+                             (1.0f / 6.0f);
```
Comment (lazarevevgeny) [inline ngraph/core/include/ngraph/runtime/reference/hswish.hpp:34]: Just divide by 6.0.

File: inference-engine/src/transformations/src/transformations/hswish_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        if (data.size() != 1 || std::fabs(data[0] - value) < std::numeric_limits<float>::epsilon()) {
+            return false;
+        }
+    } else {
+        return false;
+    }
+    return true;
+}
+
+ngraph::pass::HSwishFusionWithRelu::HSwishFusionWithRelu() {
```
Comment (lazarevevgeny) [inline inference-engine/src/transformations/src/transformations/hswish_fusion.cpp:28]: Add this transformation to inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp

File: ngraph/core/include/ngraph/opsets/opset4_tbl.hpp [hunk 3]
```diff
@@ -62,6 +62,7 @@ NGRAPH_OP(GroupConvolution, ngraph::op::v1)
 NGRAPH_OP(GroupConvolutionBackpropData, ngraph::op::v1)
 NGRAPH_OP(GRN, ngraph::op::v0)
 NGRAPH_OP(HardSigmoid, ngraph::op::v0)
+NGRAPH_OP(HSwish, ngraph::op::v4)
```
Comment (lazarevevgeny) [inline ngraph/core/include/ngraph/opsets/opset4_tbl.hpp:65]: Please, move this line to the end of the file with section of opset4 ops.

File: ngraph/core/include/ngraph/runtime/reference/hswish.hpp (hunk 1 above)
Comment (mitruska) [reply ngraph/core/include/ngraph/runtime/reference/hswish.hpp]: Don't you think that multiplication can be faster?

Comment (lazarevevgeny) [reply ngraph/core/include/ngraph/runtime/reference/hswish.hpp]: Yes, the multiplication will be faster. But the code will look more readable for the end user. And I am pretty sure that all modern compilers convert division with constant to a multiplication with `1 / const` internally.

Comment (mitruska) [reply ngraph/core/include/ngraph/runtime/reference/hswish.hpp]: Ok, done

File: ngraph/core/include/ngraph/op/hswish.hpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+{
+    namespace op
+    {
+        namespace v4
+        {
+            /// \brief A HSwish Activation Function
+            /// f(x) =  x * (min(max(x + 3, 0), 6) / 6) or
+            /// f(x) = x * (min(ReLU(x + 3), 6) / 6)
+            ///
+            class NGRAPH_API HSwish : public ngraph::op::Op
```
Comment (lazarevevgeny) [inline ngraph/core/include/ngraph/op/hswish.hpp:32]: Please, add the operation to the Python API.

**Comment by lazarevevgeny** [comment]
@mitruska, please, merge the recent master and convert the PR from draft to a regular one.

File: inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+
+ngraph::pass::HSwishDecomposition::HSwishDecomposition() {
+    // Decomposes HSwish(x) op into sub-graph x * (min(Relu(x + 3), 6) * const(1/6)
+    auto hswish = ngraph::pattern::wrap_type<opset4::HSwish>();
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) {
+        auto &pattern_to_output = m.get_pattern_value_map();
+        auto hswish_node =  std::dynamic_pointer_cast<ngraph::opset4::HSwish>(pattern_to_output.at(hswish).get_node_shared_ptr());
```
Comment (lazarevevgeny) [inline inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp:19]: redundant space character.

File: inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        auto input_type = hswish_node->input_value(0).get_element_type();
+        auto add_constant = ngraph::opset4::Constant::create(input_type, ngraph::Shape{}, {3.0});
+        auto add = std::make_shared<ngraph::opset4::Add>(hswish_node->input_value(0), add_constant);
+        auto relu = std::make_shared<ngraph::opset4::Relu>(add);
+        auto min_constant = ngraph::opset4::Constant::create(input_type, ngraph::Shape{}, {6.0});
+        auto min = std::make_shared<ngraph::opset4::Minimum>(relu, min_constant);
+        auto mul_first = std::make_shared<ngraph::opset4::Multiply>(hswish_node->input_value(0), min);
+        auto mul_constant = ngraph::opset4::Constant::create(input_type, ngraph::Shape{}, {0.1666666716});  // const(1/6)
```
Comment (lazarevevgeny) [inline inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp:32]: May be write `1.0 / 6.0`?

File: inference-engine/tests/functional/inference_engine/transformations/hswish_fusion_test.cpp [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        ASSERT_NO_THROW(check_rt_info(f));
+    }
+
+    {
+        auto input = std::make_shared<ngraph::opset4::Parameter>(ngraph::element::f32, ngraph::Shape{});
+        auto hswish = std::make_shared<ngraph::opset4::HSwish>(input);
+
+        f_ref = std::make_shared<ngraph::Function>(ngraph::NodeVector{hswish}, ngraph::ParameterVector{input});
+    }
+
```
Comment (lazarevevgeny) [inline inference-engine/tests/functional/inference_engine/transformations/hswish_fusion_test.cpp:82]: too many empty lines

File: ngraph/python/src/ngraph/opset4/__init__.py [hunk 8]
```diff
@@ -72,6 +72,7 @@
 from ngraph.opset1.ops import group_convolution_backprop_data
 from ngraph.opset3.ops import gru_cell
 from ngraph.opset1.ops import hard_sigmoid
+from ngraph.opset1.ops import hswish
```
Comment (lazarevevgeny) [inline ngraph/python/src/ngraph/opset4/__init__.py:75]: `opset1` -> `opset4`

File: inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp (hunk 6 above)
Comment (mitruska) [reply inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp]: Ok, I was wondering what is more stable.

File: ngraph/core/src/op/hswish.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    bool evaluate_hswish(const HostTensorPtr& arg, const HostTensorPtr& out, const size_t count)
+    {
+        bool rc = true;
+        out->set_unary(arg);
+
+        switch (arg->get_element_type())
+        {
+            TYPE_CASE(f16)(arg, out, count);
```
Comment (lazarevevgeny) [inline ngraph/core/src/op/hswish.cpp:69]: Please, add `bf16`

File: ngraph/core/include/ngraph/runtime/reference/hswish.hpp [hunk 10]
```diff
@@ -0,0 +1,38 @@
+//*****************************************************************************
```
Comment (ilyachur) [inline ngraph/core/include/ngraph/runtime/reference/hswish.hpp:1]: Please move reference implementation to `ngraph/core/reference`

File: ngraph/core/include/ngraph/op/hswish.hpp [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+{
+    namespace op
+    {
+        namespace v4
+        {
+            /// \brief A HSwish Activation Function
+            /// f(x) =  x * min(max(x + 3, 0), 6) / 6 or
+            /// f(x) = x * min(ReLU(x + 3), 6) / 6
+            ///
+            class NGRAPH_API HSwish : public ngraph::op::Op
```
Comment (mbencer) [inline ngraph/core/include/ngraph/op/hswish.hpp:32]: IMO `HSwish` can inherit from `ngraph::op::util::UnaryElementwiseArithmetic`. It provides some additional checks (for example block user to pass boolean element type) and provides shape inference (validate_and_infer_types will be not needed).
Looks for example for `Relu`.

Comment (mitruska) [reply ngraph/core/include/ngraph/op/hswish.hpp]: I thought it's deprecated, but seems legit. 
Thanks, done.

File: ngraph/core/src/op/hswish.cpp (hunk 9 above)
Comment (mitruska) [reply ngraph/core/src/op/hswish.cpp:65]: Done

File: ngraph/core/include/ngraph/runtime/reference/hswish.hpp (hunk 10 above)
Comment (mitruska) [reply ngraph/core/include/ngraph/runtime/reference/hswish.hpp]: Done

File: inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+ngraph::pass::HSwishDecomposition::HSwishDecomposition() {
+    // Decomposes HSwish(x) op into sub-graph x * (min(Relu(x + 3), 6) * const(1/6)
+    auto hswish = ngraph::pattern::wrap_type<opset4::HSwish>();
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) {
+        auto &pattern_to_output = m.get_pattern_value_map();
+        auto hswish_node = std::dynamic_pointer_cast<ngraph::opset4::HSwish>(pattern_to_output.at(hswish).get_node_shared_ptr());
+
+        if (m_transformation_callback(hswish_node)) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp:21]: Please check that `hswish_node` is not nullptr here

File: inference-engine/tests/functional/inference_engine/transformations/hswish_fusion_test.cpp [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+        auto min = std::make_shared<ngraph::opset4::Minimum>(relu, min_constant);
+        auto mul = std::make_shared<ngraph::opset4::Multiply>(input, min);
+        auto div_constant = ngraph::opset4::Constant::create(ngraph::element::f16, ngraph::Shape{}, {6.0});
+        auto div = std::make_shared<ngraph::opset4::Divide>(mul, div_constant);
+
+        f = std::make_shared<ngraph::Function>(ngraph::NodeVector{div}, ngraph::ParameterVector{input});
+
+        ngraph::pass::Manager manager;
+        manager.register_pass<ngraph::pass::InitNodeInfo>();
+        manager.register_pass<ngraph::pass::HSwishFusion>();
```
Comment (GlebKazantaev) [inline inference-engine/tests/functional/inference_engine/transformations/hswish_fusion_test.cpp:38]: I think it's better to register particular MatcherPass that is expected to be executed on this sub-graph (the same comment to other fusion tests).

File: ngraph/core/include/ngraph/op/hswish.hpp [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+        namespace v4
+        {
+            /// \brief A HSwish Activation Function
+            /// f(x) =  x * min(max(x + 3, 0), 6) / 6 or
+            /// f(x) = x * min(ReLU(x + 3), 6) / 6
+            ///
+            class NGRAPH_API HSwish : public ngraph::op::util::UnaryElementwiseArithmetic
+            {
+            public:
+                static constexpr NodeTypeInfo type_info{"HSwish", 4};
```
Comment (GlebKazantaev) [inline ngraph/core/include/ngraph/op/hswish.hpp:36]: Please use NGRAPH_RTTI_DECLARATION/DEFINITION

**Comment by GlebKazantaev** [comment]
@mitruska please rebase to the latest master.

File: ngraph/core/include/ngraph/op/hswish.hpp (hunk 14 above)
Comment (GlebKazantaev) [reply ngraph/core/include/ngraph/op/hswish.hpp:36]: In hpp file: `NGRAPH_RTTI_DECLARATION;`
In cpp file: `NGRAPH_RTTI_DEFINITION(op::v4::HSwish , "HSwish", 4);`

File: inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+
+ngraph::pass::HSwishDecomposition::HSwishDecomposition() {
+    // Decomposes HSwish(x) op into sub-graph x * (min(Relu(x + 3), 6) * const(1/6)
+    auto hswish = ngraph::pattern::wrap_type<opset4::HSwish>();
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) {
+        auto &pattern_to_output = m.get_pattern_value_map();
+        auto hswish_node = std::dynamic_pointer_cast<ngraph::opset4::HSwish>(pattern_to_output.at(hswish).get_node_shared_ptr());
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp:19]: BTW you don't need to cast it to HSwish as you use only its inputs

File: inference-engine/src/transformations/src/transformations/hswish_fusion.cpp [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+
+#include "transformations/hswish_fusion.hpp"
+
+#include <memory>
+
+#include <ngraph/opsets/opset4.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+
+bool check_constant_value(const std::shared_ptr<ngraph::opset4::Constant>& constant,
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/hswish_fusion.cpp:13]: Can we unify this function and function  from `swish_fusion.cpp` and move it to utils or some header?

Comment (postrational) [reply inference-engine/src/transformations/src/transformations/hswish_fusion.cpp:13]: If we move this function to utils, then we should make sure it has a unique and self-documenting name.
Something like `is_constant_float` or something similar...

File: ngraph/python/tests/test_ngraph/test_hswish.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
+# ******************************************************************************
```
Comment (postrational) [inline ngraph/python/tests/test_ngraph/test_hswish.py:1]: Perhaps this test could be moved to a file with all other activation functions?

File: inference-engine/src/transformations/src/transformations/hswish_fusion.cpp (hunk 16 above)
Comment (mitruska) [reply inference-engine/src/transformations/src/transformations/hswish_fusion.cpp:13]: I also thought about it, but I wouldn't like to touch the Swish op files in this PR. I can create a separate PR with enabling this function as a helper, after this one will be merged.

File: ngraph/python/tests/test_ngraph/test_hswish.py (hunk 17 above)
Comment (mitruska) [reply ngraph/python/tests/test_ngraph/test_hswish.py:1]: I couldn't find such a file, do you know the path?

File: inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp (hunk 15 above)
Comment (mitruska) [reply inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp]: Changed

File: ngraph/core/include/ngraph/op/hswish.hpp (hunk 14 above)
Comment (mitruska) [reply ngraph/core/include/ngraph/op/hswish.hpp]: Done

File: inference-engine/tests/functional/inference_engine/transformations/hswish_fusion_test.cpp (hunk 13 above)
Comment (mitruska) [reply inference-engine/tests/functional/inference_engine/transformations/hswish_fusion_test.cpp]: Make sense, done

File: inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp (hunk 12 above)
Comment (mitruska) [reply inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp]: Done

File: ngraph/python/tests/test_ngraph/test_hswish.py (hunk 17 above)
Comment (mitruska) [reply ngraph/python/tests/test_ngraph/test_hswish.py]: I moved the test to `ngraph/python/tests/test_ngraph/test_ops_unary.py`

File: inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+ngraph::pass::HSwishDecomposition::HSwishDecomposition() {
+    // Decomposes HSwish(x) op into sub-graph x * (min(Relu(x + 3), 6) * const(1/6)
+    auto hswish = ngraph::pattern::wrap_type<opset4::HSwish>();
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher &m) {
+        auto &pattern_to_output = m.get_pattern_value_map();
+        auto hswish_node = pattern_to_output.at(hswish).get_node_shared_ptr();
+
+        if (hswish_node == nullptr || m_transformation_callback(hswish_node)) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp:21]: nit: not need to check for nullptr as you don't have dynamic cast for this ptr

File: inference-engine/src/transformations/src/transformations/hswish_fusion.cpp (hunk 16 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/hswish_fusion.cpp:13]: @mitruska ok

File: inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp (hunk 18 above)
Comment (mitruska) [reply inference-engine/src/transformations/src/transformations/hswish_decomposition.cpp]: Right

**Pull request merged**